core/ir/synthetic/encode.c

```diff
@@ -0,0 +1,129 @@
+     * values.
+     */
+    uint encoding_header = 0;
+
+    /* Encode number of register destination operands (i.e., written registers).
+     */
+    uint num_dsts = (uint)instr_num_dsts(instr);
+    encoding_header |= num_dsts;
+
+    /* Encode number of register source operands (i.e., read registers).
+     */
+    uint num_srcs = (uint)instr_num_srcs(instr);
+    encoding_header |= (num_srcs << SRC_OPND_SHIFT);
+
+    /* Encode arithmetic flags.
+     */
+    uint eflags_instr = instr_get_arith_flags(instr, DR_QUERY_DEFAULT);
+    uint eflags = 0;
+    if (TESTANY(EFLAGS_WRITE_ARITH, eflags_instr))
+        eflags |= SYNTHETIC_INSTR_WRITES_ARITH;
+    if (TESTANY(EFLAGS_READ_ARITH, eflags_instr))
+        eflags |= SYNTHETIC_INSTR_READS_ARITH;
+    encoding_header |= (eflags << FLAGS_SHIFT);
+
+    /* Encode category as synthetic opcode.
+     */
+    uint category = instr_get_category(instr);
+    encoding_header |= (category << CATEGORY_SHIFT);
+
+    /* Copy header encoding back into encoded_instr output.
+     */
+    *((uint *)&encoded_instr[0]) = encoding_header;
+
+    /* Encode register destination operands, if present.
+     */
+    opnd_size_t max_opnd_size = OPSZ_NA;
+    for (uint dst_index = 0; dst_index < num_dsts; ++dst_index) {
+        opnd_t dst_opnd = instr_get_dst(instr, dst_index);
+        max_opnd_size = opnd_get_size(dst_opnd);
+        uint num_regs_used_by_opnd = (uint)opnd_num_regs_used(dst_opnd);
+        for (uint opnd_index = 0; opnd_index < num_regs_used_by_opnd; ++opnd_index) {
+            reg_id_t reg = opnd_get_reg_used(dst_opnd, opnd_index);
+            encoded_instr[dst_index + OPND_INDEX] = (byte)reg;
+        }
+    }
+
+    /* Encode register source operands, if present.
+     */
+    for (uint src_index = 0; src_index < num_srcs; ++src_index) {
+        opnd_t src_opnd = instr_get_src(instr, src_index);
+        max_opnd_size = opnd_get_size(src_opnd);
+        uint num_regs_used_by_opnd = (uint)opnd_num_regs_used(src_opnd);
+        for (uint opnd_index = 0; opnd_index < num_regs_used_by_opnd; ++opnd_index) {
+            reg_id_t reg = opnd_get_reg_used(src_opnd, opnd_index);
+            encoded_instr[src_index + OPND_INDEX + num_dsts] = (byte)reg;
+        }
+    }
+
+    /* Encode largest register size, if there is at least one operand.
+     */
+    uint num_opnds = num_dsts + num_srcs;
+    if (num_opnds > 0) {
+        CLIENT_ASSERT(
+            max_opnd_size != OPSZ_NA,
+            "instructions with register operands cannot have operand size OPSZ_NA");
+        encoded_instr[OP_SIZE_INDEX] = (byte)max_opnd_size;
+    }
+
+    /* Retrieve instruction length, which includes bytes for padding to reach 4 byte
+     * alignment.
+     * This already accounts for 1 additional byte containing max register operand size,
+     * if there are any operands.
+     */
+    uint length = (uint)instr_length(dcontext, instr);
```

Comment:
Will this go do a decode?  Shouldn't we compute it directly ourselves?

Reply:
We can, but the assumption was that a REGDEPS instr_t is always fully decoded.

Reply:
This worked because of `instr_set_raw_bits_valid(instr, true)`.
Now that we removed it, we compute it.